File: implementations/java/client/src/main/java/com/theshuai/tunnelclient/peer/PeerUdpProbeCodec.java
package com.theshuai.tunnelclient.peer;

import com.theshuai.common.peermesh.PeerUdpProbe;
import com.theshuai.common.util.JsonUtil;

import java.nio.charset.StandardCharsets;

final class PeerUdpProbeCodec {
    static final int MAX_PACKET_BYTES = 2_048;
    private static final byte[] MAGIC = PeerUdpProbe.MAGIC.getBytes(StandardCharsets.US_ASCII);

    private PeerUdpProbeCodec() {
    }

    static PeerUdpProbe decode(byte[] packet, int offset, int length) {
        if (!looksPlausible(packet, offset, length)) {
            return null;
        }
        PeerUdpProbe probe = JsonUtil.bytesToObjectQuietly(packet, offset, length, PeerUdpProbe.class);
        return probe != null && PeerUdpProbe.MAGIC.equals(probe.getMagic()) ? probe : null;
    }

    private static boolean looksPlausible(byte[] packet, int offset, int length) {
        if (packet == null
                || offset < 0
                || length < MAGIC.length + 8
                || length > MAX_PACKET_BYTES
                || offset > packet.length - length
                || packet[offset] != '{'
                || packet[offset + length - 1] != '}') {
            return false;
        }
        int searchEnd = Math.min(offset + length, offset + 160);
        outer:
        for (int index = offset; index <= searchEnd - MAGIC.length; index++) {
            for (int magicIndex = 0; magicIndex < MAGIC.length; magicIndex++) {
                if (packet[index + magicIndex] != MAGIC[magicIndex]) {
                    continue outer;
                }
            }
            return true;
        }
        return false;
    }
}

File: implementations/csharp/client/src/Specus.Client/PeerMesh/PeerUdpProbeCodec.cs
using System.Text;
using System.Text.Json;

namespace Specus.Client.PeerMesh;

/// <summary>Bounded, exception-free decoder for unauthenticated UDP connectivity probes.</summary>
internal static class PeerUdpProbeCodec
{
    internal const int MaxPacketBytes = 2_048;
    internal const string Magic = "specus-peer-mesh";

    private static readonly byte[] MagicBytes = Encoding.ASCII.GetBytes(Magic);
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);

    internal static PeerMeshClient.PeerUdpProbe? Decode(ReadOnlySpan<byte> packet)
    {
        if (!LooksPlausible(packet))
        {
            return null;
        }
        try
        {
            var probe = JsonSerializer.Deserialize<PeerMeshClient.PeerUdpProbe>(packet, JsonOptions);
            return probe is not null && string.Equals(probe.Magic, Magic, StringComparison.Ordinal)
                ? probe
                : null;
        }
        catch (Exception ex) when (ex is JsonException or NotSupportedException)
        {
            return null;
        }
    }

    private static bool LooksPlausible(ReadOnlySpan<byte> packet)
    {
        if (packet.Length < MagicBytes.Length + 8
            || packet.Length > MaxPacketBytes
            || packet[0] != (byte)'{'
            || packet[^1] != (byte)'}')
        {
            return false;
        }
        var searchLength = Math.Min(packet.Length, 160);
        return packet[..searchLength].IndexOf(MagicBytes) >= 0;
    }
}
